### PR TITLE
Merge (#1)

### LINKED ISSUE #36934: For loop variable 'tensor_value' is not used in the loop body
Hello, I'm curious about this. The for loop I mentioned is using variable `tensor_values`, but the variable `tensor_value` is never be used in the loop itself. Is it necessary?

https://github.com/tensorflow/tensorflow/blob/aadf705c858014a168d1a582accc81b7cc774d68/tensorflow/python/debug/lib/dumping_callback_test.py#L1131-L1136

### PR DESCRIPTION
**Pull request by jaketae**
* Raise error if we hit an unknown exception while reading file from s3

* Add forbidden error check

* Added relevant changes for retrying operations in the s3 file system
Add header as in master for building the test, and change error type for delete dir
Register retrying file system for s3
Compile retrying s3 file system
Add forbidden errors, add mtime to the stat function, add if check around creating file in createDir
Move read error bug to separate PR for easier reviews

Lint fix with buildifier

* Refactored error handling to separate methods for reuse

* Make function static

* Fix link failure for tests

* update ragged_string_ops.py comment

* [Intel Mkl] Upgrade Sqlite3 to fix CVE-2019-19880 CVE-2019-19244 and CVE-2019-19645

* Use CreateStatusFromAwsError for the read error

* make gpu delegate build

it seems this line was forgotten when adding space_to_depth

* Depthwise convolution 3x3 per-channel int8 for dot-product ARM (17x).

Disable dot-product path temporarily.

PiperOrigin-RevId: 297134519
Change-Id: Ib00acb656471143fd31ca39e6762d46e9dec994a

* Add method that returns the version for hexagon_interface.
This will be used to compare version between hexagon_interface and hexagon_skel libraries.

PiperOrigin-RevId: 297138592
Change-Id: I23e1279fa5baa9cb262308f00d5f7f55f1256221

* Use strategy properties to check if a worker should checkpoint instead of use the TF_CONFIG. Using TF_CONFIG will cause errors if we use a cluster resolver to initialize the strategy.

PiperOrigin-RevId: 297139161
Change-Id: I7f13c7b7ea6588e020011569b26eb0bf49893bbf

* Enabling non trivial fast tuning for all vendors.

PiperOrigin-RevId: 297142417
Change-Id: I948e0ae458fbf59841d83d3c087c35c286daf269

* Update Eigen to:
https://gitlab.com/libeigen/eigen/-/commit/52a2fbbb008a47c5e3fb8ac1c65c2feecb0c511c

PiperOrigin-RevId: 297148140
Change-Id: I71fa876fcf492dc0051be3c3eecd12407770220c

* Qualify uses of std::string

PiperOrigin-RevId: 297148457
Change-Id: Id91f1da7d30e9a9880d8610c36ef8657f12817f0

* Improving snapshot read performance (under snappy compression) by reducing the
number of copies by 2.

We do the following in this CL.

1) Get rid of the snappy input and output buffers that were previously being
used to do the compression. This saved one copy.
2) Directly decompress the compressed bytes into the TensorBuffer for simple
types (not string, variant, resource). This saves another copy during Tensor
creation. For complex types, we still continue to use the TensorProto encoding
and pay a copy there.
3) As a result, we end up changing the on-disk format for Snapshot. For a group
of tensors that make up one element of an IteratorGetNext output, we first
write out a metadata proto that describes the types, shapes and sizes of
tensors. After that we lay out the Tensor data (TensorBuffers for simple types
and TensorProtos serialized for complex ones) and compress them via snappy.
4) Add a version to the SnapshotMetadata. If it isn't set its assumed to be 0
and the old code path runs. We now set it to 1 while writing so that all new
snapshots are written in this data format.

PiperOrigin-RevId: 297149479
Change-Id: I2c9a35c5a254189a5fad946b2995f25cdc452308

* update_bazel_macos is not used any more and has been replaced with install_bazelisk

PiperOrigin-RevId: 297149906
Change-Id: Ibcc1a114468a57cdf5314a7499e5a4ad32b8e941

* Replaced tools with exec_tools in genrules.

PiperOrigin-RevId: 297150637
Change-Id: I4b807e07a6b9072d53191ee7c2c3634cc0ca8e98

* Check for TENSORFLOW_LITE_PROTOS instead of __ANDROID__ when checking whether the full/lite protobuf implementation is being used.

PiperOrigin-RevId: 297150726
Change-Id: I5b7930811ef4410bfb85465736749558d6d630f1

* [tfdbg2] Fix a copy-paste error in unit test code

Fixes https://github.com/tensorflow/tensorflow/issues/36934

PiperOrigin-RevId: 297162227
Change-Id: Idcc10fdf84cfeabd6f136725923ef39ea27c4e04

* Fix an issue with out of order execution on a corner cases when there is multiple remote pending requests and remote functions with remote inputs.

Specifically:
1. We keep track the last device used to execute a remote function with remote inputs
2. Before executing a remote function, if the remote device is different than the last tracked device, we insert a sync point.
PiperOrigin-RevId: 297163787
Change-Id: I944669b92dd1b50cdf213c73e1cdf990c9011e00

* Fix the deprecation_test for py3.8

PiperOrigin-RevId: 297168125
Change-Id: I441c6cd7d85fc8e8f8aedb53d1b2ec69594d9bcd

* Add tf.MatrixBandPartOp ODS definition and corresponding verifier.

PiperOrigin-RevId: 297174451
Change-Id: I77400245d526dc647dd17f532b935bf6f942db62

* Fix device placement logic in ConvertToEagerTensor.

The fix consists in always creating a host tensor: the inputs to
ConvertToEagerTensor are host Python objects, so it makes sense that
the created tensor should be a host tensor too. The user can control
GPU copies by using tf.identity.

PiperOrigin-RevId: 297174740
Change-Id: I01f2aa9be3eb29fd49c7d81823e044db292b2d7c

* *In dilated conv pass, set shape for expand_dims result correctly.
*Put more checks on input conv2d's dilation attributes, padding type, etc.

PiperOrigin-RevId: 297174875
Change-Id: Ibc59cc80a091e98d3eb1f7a855a4ee08402d7d96

* Experimental prototype of BertTokenizer  for tflite support library

PiperOrigin-RevId: 297176134
Change-Id: I4fea84735d1686f1ae1e98f4433bd0e29a642d35

* Remove ops.device(None).__enter__ from eager:core_test.

Those statements are not necessary for the test to consistently pass.

PiperOrigin-RevId: 297180465
Change-Id: I7dd6798b56fa7f80b8170ddc92ed811370bffd2f

* Bump open source llvm revision to fee41517fe0f7ff9f0e204dd9200ebf32ca03cb8

PiperOrigin-RevId: 297184332
Change-Id: Ide9c91a6171350d9eadb406f3ed028bcdbc33cc9

* Add `shape_signature` to visualize.py.

PiperOrigin-RevId: 297194460
Change-Id: I9b08d418bf75b6b5f4e2e2f7b05c84932d8bab46

* Expose quantization parameters in the TFLite Java API

PiperOrigin-RevId: 297207151
Change-Id: Iff5a1e720df750d0c90710e5209bcabc53054150

* Go: Update generated wrapper functions for TensorFlow ops.

PiperOrigin-RevId: 297207526
Change-Id: I92b95973a96d2c8641b5c13d4c8addf7687df1c9

* Update hexagon_nn_headers to v1.10.3.1.3
Changes Includes:
* Support soc_id:371
* New method exposed that returns the version of hexagon_nn used in libhexagon_interface.so

PiperOrigin-RevId: 297212018
Change-Id: I5c3396f04b305ea6197e8c693c5cd06c789e8d8d

* Qualify uses of std::string

PiperOrigin-RevId: 297212802
Change-Id: Ic65150e7ab418be034f48d45ce25ef5d19105836

* Internal refactoring for metrics.

PiperOrigin-RevId: 297214293
Change-Id: I9ef1510677669bc9c21de5839fde7ed42421fb6d

* Add test for connecting to v2-32 Cloud TPU.

PiperOrigin-RevId: 297219348
Change-Id: I6072853736fc98badbca598850f850b9980cc7b0

* Fix Concat behavior in Inception V3 quant

PiperOrigin-RevId: 297220987
Change-Id: Icd045f9dac3a7cd6aa42bf5070c206bc7b375046

* Minor change in update script

PiperOrigin-RevId: 297221831
Change-Id: Ibc864c21ad7bd277ffe00caecb9751629f89d97b

* Extend the optimize global tensors pass with resource analysis

Inter-procedural resource analysis for identifying resource as mutable or not.

PiperOrigin-RevId: 297225621
Change-Id: I12690ddab584660792c216347efa7a265fa6ffee

* Add check for hexagon version between interface and skel.
libhexagon_interface and libhexagon_nn_skel should be used with matching versions. Users shouldn't use mixed versions as it will not work correctly.

PiperOrigin-RevId: 297239641
Change-Id: I4e82626fc395eaba23fd51f84fbb402e24b2fd8a

* Improves documentation for keras.preprocessing.pad_sequence

PiperOrigin-RevId: 297247197
Change-Id: I4c85e8ba6d4ae43d4c249442ef9c47bb6a5805c6

* Go: Update generated wrapper functions for TensorFlow ops.

PiperOrigin-RevId: 297248087
Change-Id: Ibc3efb37f0f855fb86f8d0915c17c5bfb26847d3

* Upgrade nanopb to latest release.

Fixes security vulnerability reported in #37011

PiperOrigin-RevId: 297248686
Change-Id: Ib24b109c8a36b8673842b2f2ebb771406e5fa022

* Split the source files into ESP IDF components by their paths.

PiperOrigin-RevId: 297249461
Change-Id: I16c81c416ab803d4ba82cc4ae801b35a00be1acc

* Special-case variable reads in auto control deps.
1. Variable reads, with no write in between, are no longer performed in a sequence. This included reads happening inside functional ops e.g. nested functions or control flow.
2. Variable reads are no longer added to the function's control outputs.

PiperOrigin-RevId: 297253117
Change-Id: I833891f644cf6f8c1b733401ed2e3de3adb9ef0e

* Add GPU kernel stats to cloud tools.

PiperOrigin-RevId: 297267405
Change-Id: Ia3f0ac52ccb1db07f0b6230f9b181e8f2e316c04

* Set model.history after `on_epoch_end` in `History` class.

PiperOrigin-RevId: 297267664
Change-Id: I0cf97412913bdfe3599d74f162d172f02e7ceac9

* Switch ConcatOp, PackOp, SelectOp, and SelectV2Op to access inputs by index (not name).

These kernels are heavily used, and the name resolution on each invocation causes non-trivial overhead.

PiperOrigin-RevId: 297267975
Change-Id: Id69de0e2cff3622e992389c16e020a5da3141462

* Add an experimental_hints to batch all reduce

This contains all performance hints to the API. Currently there's only bytes_per_pack, which splits large batches into multiple packs allows overlapping communication and computation.

Currently we can only pack if all Tensors in the batch have known shapes.

PiperOrigin-RevId: 297269428
Change-Id: Iaf7d7d3adf7c6cad59aa6079fbcd36b31e92c4b5

* Set resource as mutable when unknown op is encountered in the optimize global tensors pass

PiperOrigin-RevId: 297274143
Change-Id: Ia33df38d23875a1a4021c52e70fca17b8aeb9c96

* [OpKernel] Implement `OpKernelContext::output_required(int)`.

This long-pending feature enables kernels to specialize for the case where not all of their outputs may need to be produced. As an example, this change uses `OpKernelContext::output_required()` in `SparseFillEmptyRowsOp`, which often produces unused outputs for the purpose of backpropagation; using `output_required()` we can save the related allocations in inference workloads.

I did consider implementing this as a Grappler rewrite pass, but the example of SparseFillEmptyRowsOp convinced me that (given the present state) a runtime test was better. The rewrite-based alternative would require me to have four similar op registrations for SparseFillEmptyRows (since we now have two optional outputs that may be present or absent, so there are 2^2 possible signatures), and to devise a mapping/registration scheme that the rewriter might use to substitute different implementations. By contrast, the runtime check is (i) pretty cheap compared to op dispatch, and (ii) easy to implement gradually and locally.

PiperOrigin-RevId: 297274989
Change-Id: I23b5207017921ba118bd4fc31dc54ff53fe4332d

* Include FreeRTOS.h before other dependencies demending this prior inclusion.

PiperOrigin-RevId: 297275456
Change-Id: I624bc648276a4cad3ca481cc2433fd29754ddfa6

* Inject keras.preprocessing modules into the doctest.

PiperOrigin-RevId: 297276571
Change-Id: I76c18e90f02ea6d0ef7ad69dae16e231734efc8d

* Fix the BatchNorm v2 performance issue wrt "training".

Change the logic to skip the math_ops.logical_and() when layer.trainable is True, since the "training" value is only going to be changed if layer.trainable is False.

This apparently fix the regression issue (1.33x). I think the root cause might be the logical_and() with a value that is not the GPU device (layer.trainable).

PiperOrigin-RevId: 297277503
Change-Id: I9e3b6650bfa58c3d2ea8e3f023a3bd1a5160cc8d

* Add warmup in keras tb callback to improve profiling accuracy.

PiperOrigin-RevId: 297277543
Change-Id: Ic78a286400fffa73d583f3bb27bfce0dc1e275dc

* Go: Update generated wrapper functions for TensorFlow ops.

PiperOrigin-RevId: 297278005
Change-Id: I7ae629b04ffb9c44c732dac221b0821f86b282e8

* Consolidate via CallOpInterface the various TF call ops handling in optimize global tensors pass

PiperOrigin-RevId: 297283013
Change-Id: Ibd5a306b662d61ef2644a6cf59f78ac0119aa02e

* Add go_backwards support for keras fused lstm

PiperOrigin-RevId: 297287353
Change-Id: Idaebe3d0c84fc8be03651233a4af7c9cd46a23ca

* Go: Update generated wrapper functions for TensorFlow ops.

PiperOrigin-RevId: 297288470
Change-Id: I86fd6955e320399697a6f1082c8a98efdd09d639

Co-authored-by: Rahul Huilgol <rahulhuilgol@gmail.com>
Co-authored-by: HUAN-PING SU <pingsutw@gmail.com>
Co-authored-by: Clayne Robison <clayne.b.robison@intel.com>
Co-authored-by: "freedom" Koan-Sin Tan <koansin.tan@gmail.com>
Co-authored-by: TensorFlower Gardener <gardener@tensorflow.org>
Co-authored-by: Alex Stark <539273+jalexstark@users.noreply.github.com>
Co-authored-by: Karim Nosseir <44206880+karimnosseir@users.noreply.github.com>
Co-authored-by: anj-s <32556631+anj-s@users.noreply.github.com>
Co-authored-by: Rohan Jain <rohan100jain@gmail.com>
Co-authored-by: Brian Atkinson <nairb774@gmail.com>
Co-authored-by: Mihai Maruseac <mihai.maruseac@gmail.com>
Co-authored-by: Shanqing Cai <cais@google.com>
Co-authored-by: Bramandia Ramadhana <bramandia@gmail.com>
Co-authored-by: lucyrfox <lucyfox@google.com>
Co-authored-by: Haoliang Zhang <haoliang@google.com>
Co-authored-by: thuang513 <556234+thuang513@users.noreply.github.com>
Co-authored-by: River Riddle <riddleriver@gmail.com>
Co-authored-by: Nupur Garg <nupurgarg@gmail.com>
Co-authored-by: lu-wang-g <47436172+lu-wang-g@users.noreply.github.com>
Co-authored-by: Revan Sopher <rsopher@gmail.com>
Co-authored-by: Sachin Joglekar <srjoglekar@google.com>
Co-authored-by: sun51 <yanh.sun@gmail.com>
Co-authored-by: Ashwin Murthy <ashwinm@google.com>
Co-authored-by: Frédéric Rechtenstein <fred.rec@gmail.com>
Co-authored-by: Saurabh Saxena <saxenasaurabh@users.noreply.github.com>
Co-authored-by: Situ Yi <60493024+yisitu@users.noreply.github.com>
Co-authored-by: Yash Katariya <yashkatariya@google.com>
Co-authored-by: Derek Murray <derek.murray@gmail.com>
Co-authored-by: Ran Chen <crccw@google.com>
Co-authored-by: Qianli Scott Zhu <scottzhu@google.com>
Co-authored-by: renjie-liu <36247193+renjie-liu@users.noreply.github.com>

### CODE REVIEW

**Comment by jaketae** [comment]
Sincere apologies for the mistake.
